_posts/2018-02-05-unrealiable-network.md

```diff
@@ -0,0 +1,111 @@
+
+# Avoid Making Network Requests as Much as You Can
+
+The best way to avoid getting failing network requests is of course to abandom making them completely.
+
+![Failing network requests meme](/img/unrealiable-network/network_requests_meme.jpg)
+
+Thought this is very efficient, it is not very practical for most of us. Majority of web and mobile apps depend on getting data from servers, and thus, making network requests is mandatory. Still, if we cannot abandom them completely, we can at least reduce the amount of them.
+
+How can we reduce the amount of network requests? First, we should analyse what kind of requests our application is making. The developer tools of [Chrome](https://developer.chrome.com/devtools) and [Firefox](https://developer.mozilla.org/fi/docs/Tools) are very practical for this analysis. On the network tab, you can see all the requests your application is making.
+
+![Chrome Network Tab](/img/unrealiable-network/chrome_network_tab.png)
+
+The way to reduce the amount of requests depends on your application. The following questions can help you to find the requests that could be improved:
+
+**Are you making two separate network requests for saving data and then getting the same data again from the server to update the view?** 
+
+Consider changing the implementation of the data saving API to return the updated data. This way, you can save the data and update the view with one request.
+
+**Are you making multiple network requests to get data for a specific view?**
+
+Consider combining these requests, i.e. making a single API method to return all the data for a specific view. The services for the requests you previously made are already there and you can use them in creating a single API method.
+
+**Is some specific request slowing down the initial render of your application?**
+
+Your users want your application to load as quickly as possible, and thus you should only make the network requests that are absolutely mandatory to get your application up and running. If the network infrastructure is unstable, getting the application to load itself can take a long time or even fail completely.
+
+# Retry with Increasing Timeout
+
+Now that we have reduced the amount of requests, it is time to make sure those requests pass to the server and back as often as possible. If we cannot make modifications to the network infrastructure, we have to make changes in the application level.
```

Comment:
This could benefit from a treatment of what lower-level retry mechanisms are in place. There's at least TCP level, which tries to provide an abstraction of a reliable connection, and the browser level, where browsers automatically retry requests (including XMLHttpRequests, i think?). 

Web app intitiated retries may be most valuable when there are some transient errors in proxies or load balancers where the HTTP requests nominally finish, but they transiently get an error code from the other end.

A downside to retries is that it takes longer to signal the user that the network is down, when there already are the 2 lower level retry mechanisms in place.

Reply:
Good suggestion. Will be added.